File: 02-Web开发/05-流量治理与限流/02-RateLimit响应头部/python/ratelimit_headers.py
"""RateLimit / RateLimit-Policy 响应头部的解析、校验与客户端决策。

权威依据：draft-ietf-httpapi-ratelimit-headers-11《RateLimit header fields for HTTP》
（2026-05，全文 67 KB 实读）与 RFC 6585 §4（429 Too Many Requests）。

字段值本身是 RFC 9651 Structured Fields 的 List of Items：
    RateLimit-Policy: "burst";q=100;w=60,"daily";q=1000;w=86400
    RateLimit:        "default";r=50;t=30
"""

from __future__ import annotations

import base64

QUOTA_UNITS = ("requests", "content-bytes", "concurrent-requests")

PROBLEM_TYPES = {
    "quota-exceeded": "https://iana.org/assignments/http-problem-types#quota-exceeded",
    "temporary-reduced-capacity":
        "https://iana.org/assignments/http-problem-types#temporary-reduced-capacity",
    "abnormal-usage-detected":
        "https://iana.org/assignments/http-problem-types#abnormal-usage-detected",
}

# draft-11 §5 各问题类型示例里出现的状态码与原因短语（原文照录，含其笔误）
PROBLEM_STATUS = {
    "quota-exceeded": (429, "Bad Request"),        # §5.1 原文如此（429 的规范短语是 Too Many Requests）
    "temporary-reduced-capacity": (503, "Server Unavailable"),
    "abnormal-usage-detected": (429, "Too Many Requests"),
}


class HeaderError(ValueError):
    """字段值不合规范。规范 §7：Malformed RateLimit header fields MUST be ignored."""


def _split_items(value: str):
    """按顶层逗号切分 SF list（引号内与字节序列内的逗号不切）。"""
    items, buf, in_str, in_b64 = [], [], False, False
    for ch in value:
        if ch == '"':
            in_str = not in_str
            buf.append(ch)
        elif ch == ":":
            in_b64 = not in_b64
            buf.append(ch)
        elif ch == "," and not in_str and not in_b64:
            items.append("".join(buf).strip())
            buf = []
        else:
            buf.append(ch)
    tail = "".join(buf).strip()
    if tail:
        items.append(tail)
    return items


def _parse_item(raw: str):
    """解析一个 SF item：裸标识符、`"string"` 或 `:bytes:`，后接 ;k=v 参数。"""
    parts = raw.split(";")
    token = parts[0].strip()
    if token.startswith('"') and token.endswith('"') and len(token) >= 2:
        name = token[1:-1]
    elif token.startswith(":") and token.endswith(":") and len(token) >= 2:
        name = token[1:-1]
    else:
        raise HeaderError("item 必须是 String 或 Byte Sequence：%r" % token)
    params = {}
    for p in parts[1:]:
        if "=" not in p:
            raise HeaderError("参数缺值：%r" % p)
        k, v = p.split("=", 1)
        params[k.strip()] = v.strip()
    return name, params


def _as_int(params, key, required=False, allow_zero=True, default=None):
    if key not in params:
        if required:
            raise HeaderError("缺少必填参数 %r" % key)
        return default
    raw = params[key]
    try:
        v = int(raw)
    except ValueError:
        raise HeaderError("参数 %r 必须是整数：%r" % (key, raw))
    if v < 0 or (v == 0 and not allow_zero):
        raise HeaderError("参数 %r 取值非法：%d" % (key, v))
    return v


def _as_bytes(params, key):
    if key not in params:
        return None
    raw = params[key]
    if not (raw.startswith(":") and raw.endswith(":")):
        raise HeaderError("参数 %r 必须是 Byte Sequence：%r" % (key, raw))
    body = raw[1:-1]
    try:
        return base64.b64decode(body + "=" * (-len(body) % 4))
    except Exception as exc:  # noqa: BLE001
        raise HeaderError("Byte Sequence 解码失败：%r" % raw) from exc


class QuotaPolicyItem:
    """RateLimit-Policy 的一个 Item：策略名 + q/qu/w/pk。"""

    def __init__(self, name, q, qu="requests", w=None, pk=None, extras=None):
        self.name = name
        self.q = q
        self.qu = qu
        self.w = w
        self.pk = pk
        self.extras = extras or {}

    @property
    def rate(self):
        """每秒配额；无 w 时不可推算（返回 None）。"""
        if self.w is None:
            return None
        return self.q / self.w


class ServiceLimitItem:
    """RateLimit 的一个 Item：策略名 + r/t/pk。"""

    def __init__(self, name, r, t=None, pk=None, extras=None):
        self.name = name
        self.r = r
        self.t = t
        self.pk = pk
        self.extras = extras or {}

    @property
    def rate(self):
        if self.t is None or self.t == 0:
            return None
        return self.r / self.t


def parse_rate_limit_policy(value: str):
    """§3：非空 List of Items；q 必填且为非负整数；w 非负**且非零**；pk 为字节序列。"""
    items = []
    for raw in _split_items(value):
        name, params = _parse_item(raw)
        q = _as_int(params, "q", required=True)
        qu = params.get("qu")
        if qu is not None:
            if not (qu.startswith('"') and qu.endswith('"')):
                raise HeaderError("qu 必须是 String：%r" % qu)
            qu = qu[1:-1]
            if qu not in QUOTA_UNITS:
                raise HeaderError("未在配额单位注册表中的 qu：%r" % qu)
        else:
            qu = "requests"
        w = _as_int(params, "w", allow_zero=False)       # §3.1.3 非负且非零
        pk = _as_bytes(params, "pk")
        items.append(QuotaPolicyItem(name, q, qu, w, pk))
    if not items:
        raise HeaderError("RateLimit-Policy 必须是非空 List")
    return items


def parse_rate_limit(value: str):
    """§4：List of Items；r 必填非负整数；t 非负整数；pk 为字节序列。"""
    items = []
    for raw in _split_items(value):
        name, params = _parse_item(raw)
        r = _as_int(params, "r", required=True)
        t = _as_int(params, "t")
        pk = _as_bytes(params, "pk")
        items.append(ServiceLimitItem(name, r, t, pk))
    if not items:
        raise HeaderError("RateLimit 必须是非空 List")
    return items


def safe_wait(headers: dict, retry_after=None):
    """客户端决策：返回应当等待的秒数。

    优先级链（§7）：Retry-After 存在时 **MUST** 优先，有效窗口 MAY 被忽略；
    否则取所有 RateLimit Item 中最保守（速率最小）的一个换算出的等待时间；
    无法推算（无 t）时返回 0——规范不要求客户端此时退避。
    """
    if retry_after is not None:
        return float(retry_after)
    raw = headers.get("RateLimit")
    if not raw:
        return 0.0
    try:
        items = parse_rate_limit(raw)
    except HeaderError:
        return 0.0                       # §7 畸形字段 MUST be ignored
    best = 0.0
    for it in items:
        if it.rate is None:
            continue
        best = max(best, 1.0 / it.rate)  # 最保守：按该速率发一个请求所需时间
    return best


def is_throttled_problem(payload: dict):
    """§5：识别三种问题类型。返回 (是否限流类问题, violated-policies)。"""
    t = payload.get("type", "")
    for key, uri in PROBLEM_TYPES.items():
        if t == uri or t.endswith("#" + key):
            return True, list(payload.get("violated-policies", []))
    return False, []

File: 02-Web开发/05-流量治理与限流/02-RateLimit响应头部/go/ratelimit_headers.go
// Package ratelimitheaders 解析 draft-ietf-httpapi-ratelimit-headers-11 定义的
// RateLimit-Policy 与 RateLimit 响应头部，并给出客户端等待决策。
package ratelimitheaders

import (
	"encoding/base64"
	"errors"
	"strconv"
	"strings"
)

// ErrMalformed 表示字段值不合规范；按 §7 客户端 MUST 忽略该字段。
var ErrMalformed = errors.New("malformed RateLimit field")

// QuotaUnits 规范定义的三种配额单位（§3.1.2）。
var QuotaUnits = map[string]bool{
	"requests":            true,
	"content-bytes":       true,
	"concurrent-requests": true,
}

// PolicyItem 对应 RateLimit-Policy 的一个 Item（§3.1）。
type PolicyItem struct {
	Name string
	Q    int
	Qu   string
	W    int
	HasW bool
	Pk   []byte
}

// LimitItem 对应 RateLimit 的一个 Item（§4.1）。
type LimitItem struct {
	Name string
	R    int
	T    int
	HasT bool
	Pk   []byte
}

// splitItems 按顶层逗号切分 SF list。
func splitItems(value string) []string {
	var items []string
	var buf []rune
	inStr, inB64 := false, false
	for _, ch := range value {
		switch {
		case ch == '"':
			inStr = !inStr
			buf = append(buf, ch)
		case ch == ':':
			inB64 = !inB64
			buf = append(buf, ch)
		case ch == ',' && !inStr && !inB64:
			items = append(items, strings.TrimSpace(string(buf)))
			buf = nil
		default:
			buf = append(buf, ch)
		}
	}
	if tail := strings.TrimSpace(string(buf)); tail != "" {
		items = append(items, tail)
	}
	return items
}

func parseItem(raw string) (string, map[string]string, error) {
	parts := strings.Split(raw, ";")
	tok := strings.TrimSpace(parts[0])
	if len(tok) < 2 {
		return "", nil, ErrMalformed
	}
	quoted := tok[0] == '"' && tok[len(tok)-1] == '"'
	byteseq := tok[0] == ':' && tok[len(tok)-1] == ':'
	if !quoted && !byteseq {
		return "", nil, ErrMalformed
	}
	params := map[string]string{}
	for _, p := range parts[1:] {
		kv := strings.SplitN(p, "=", 2)
		if len(kv) != 2 {
			return "", nil, ErrMalformed
		}
		params[strings.TrimSpace(kv[0])] = strings.TrimSpace(kv[1])
	}
	return tok[1 : len(tok)-1], params, nil
}

func asInt(params map[string]string, key string, required, allowZero bool) (int, bool, error) {
	raw, present := params[key]
	if !present {
		if required {
			return 0, false, ErrMalformed
		}
		return 0, false, nil
	}
	v, err := strconv.Atoi(raw)
	if err != nil {
		return 0, false, ErrMalformed
	}
	if v < 0 || (v == 0 && !allowZero) {
		return 0, false, ErrMalformed
	}
	return v, true, nil
}

func asBytes(params map[string]string, key string) ([]byte, error) {
	raw, present := params[key]
	if !present {
		return nil, nil
	}
	if len(raw) < 2 || raw[0] != ':' || raw[len(raw)-1] != ':' {
		return nil, ErrMalformed
	}
	body := raw[1 : len(raw)-1]
	return base64.StdEncoding.DecodeString(body + strings.Repeat("=", -len(body)%4))
}

// ParsePolicy 解析 RateLimit-Policy（§3）。q 必填；w 非负且非零。
func ParsePolicy(value string) ([]PolicyItem, error) {
	raws := splitItems(value)
	if len(raws) == 0 {
		return nil, ErrMalformed
	}
	out := make([]PolicyItem, 0, len(raws))
	for _, raw := range raws {
		name, params, err := parseItem(raw)
		if err != nil {
			return nil, err
		}
		q, _, err := asInt(params, "q", true, true)
		if err != nil {
			return nil, err
		}
		qu := "requests"
		if v, ok := params["qu"]; ok {
			if len(v) < 2 || v[0] != '"' || v[len(v)-1] != '"' {
				return nil, ErrMalformed
			}
			qu = v[1 : len(v)-1]
			if !QuotaUnits[qu] {
				return nil, ErrMalformed
			}
		}
		w, hasW, err := asInt(params, "w", false, false)
		if err != nil {
			return nil, err
		}
		pk, err := asBytes(params, "pk")
		if err != nil {
			return nil, err
		}
		out = append(out, PolicyItem{Name: name, Q: q, Qu: qu, W: w, HasW: hasW, Pk: pk})
	}
	return out, nil
}

// ParseLimit 解析 RateLimit（§4）。r 必填；t 为非负整数（0 合法）。
func ParseLimit(value string) ([]LimitItem, error) {
	raws := splitItems(value)
	if len(raws) == 0 {
		return nil, ErrMalformed
	}
	out := make([]LimitItem, 0, len(raws))
	for _, raw := range raws {
		name, params, err := parseItem(raw)
		if err != nil {
			return nil, err
		}
		r, _, err := asInt(params, "r", true, true)
		if err != nil {
			return nil, err
		}
		t, hasT, err := asInt(params, "t", false, true)
		if err != nil {
			return nil, err
		}
		pk, err := asBytes(params, "pk")
		if err != nil {
			return nil, err
		}
		out = append(out, LimitItem{Name: name, R: r, T: t, HasT: hasT, Pk: pk})
	}
	return out, nil
}

// Rate 返回该 Item 的每秒配额；缺窗口时不可推算（返回 ok=false）。
func (l LimitItem) Rate() (float64, bool) {
	if !l.HasT || l.T == 0 {
		return 0, false
	}
	return float64(l.R) / float64(l.T), true
}

// SafeWait 客户端等待决策：Retry-After 优先，否则取所有 Item 中最保守的一个。
// 畸形字段按 §7 忽略，返回 0。
func SafeWait(rateLimit string, retryAfter float64, hasRetryAfter bool) float64 {
	if hasRetryAfter {
		return retryAfter
	}
	items, err := ParseLimit(rateLimit)
	if err != nil {
		return 0
	}
	best := 0.0
	for _, it := range items {
		r, ok := it.Rate()
		if !ok || r <= 0 {
			continue
		}
		if w := 1.0 / r; w > best {
			best = w
		}
	}
	return best
}
